jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/wih/NotifyParentCaseEventListener.java

```diff
@@ -0,0 +1,123 @@
+        notifyParentOnCompletion(event);
+    }
+
+    @Override
+    public void afterCaseCancelled(CaseCancelEvent event) {
+        notifyParentOnCompletion(event);
+    }
+
+    @Override
+    public void afterCaseDestroyed(CaseDestroyEvent event) {
+        notifyParentOnCompletion(event);
+    }
+    
+    @Override
+    public void afterProcessCompleted(ProcessCompletedEvent event) {
+        
+        CaseFileInstance caseFile = getCaseFile((KieSession) event.getKieRuntime());
+        if (caseFile != null) {
+            String caseId = ((WorkflowProcessInstanceImpl) event.getProcessInstance()).getCorrelationKey();
+            if (caseFile.getCaseId().equals(caseId)) {
+                logger.debug("Process instance {} that represents main case instance {} has completed/was aborted, notify parent if exists",
+                             event.getProcessInstance().getId(), caseId);
+                CaseEvent caseEvent = null;
+                if (event.getProcessInstance().getState() == ProcessInstance.STATE_COMPLETED) {
+                    caseEvent = new CaseCloseEvent(identityProvider.getName(), caseId, caseFile, "");
+                } else {
+                    caseEvent = new CaseCancelEvent(identityProvider.getName(), caseId, caseFile, Arrays.asList(event.getProcessInstance().getId()));
+                }
+                
+                notifyParentOnCompletion(caseEvent);
+            }
+        }
+    }
+
+    protected void notifyParentOnCompletion(CaseEvent event) {
+        CaseFileInstanceImpl caseFileInstance = (CaseFileInstanceImpl) event.getCaseFile();
+        if (caseFileInstance == null) {
+            return;
+        }
+        if (caseFileInstance.getParentInstanceId() != null && caseFileInstance.getParentWorkItemId() != null) {
+            logger.debug("Case {} has defined parent information instance {}, work item {}, going to notify it", event.getCaseId(), caseFileInstance.getParentInstanceId(), caseFileInstance.getParentWorkItemId());
+            ProcessService processService = (ProcessService) ServiceRegistry.get().service(ServiceRegistry.PROCESS_SERVICE);
+            
+            Map<String, Object> results = new HashMap<>(caseFileInstance.getData());
+            results.put("CaseId", event.getCaseId());
+            processService.completeWorkItem(caseFileInstance.getParentWorkItemId(), results);
+            logger.debug("Parent instance id {}, work item id {}, has been successfully notified about case {} completion", caseFileInstance.getParentInstanceId(), caseFileInstance.getParentWorkItemId(), event.getCaseId());
+        }
+    }
+    
+    protected CaseFileInstance getCaseFile(KieSession ksession) {
+        Collection<? extends Object> caseFiles = ksession.getObjects(new ClassObjectFilter(CaseFileInstance.class));
+        if (caseFiles.size() == 0) {
+            return null;
+        }
+        CaseFileInstance caseFile = (CaseFileInstance) caseFiles.iterator().next(); 
```

Comment:
Again, if the aim is to get a single item from a collection, because the collection should always have at most 1 item, I would rather throw an exception in case there are more items.